specs/IFrameMemoryUsage.md

```diff
@@ -0,0 +1,306 @@
+APIs for first-level iframes only. For instance, we raise the FrameCreated 
+event for first-level iframes, but not for nested iframes with depth greater than 1. 
+Also, WebView2 lacks an API to get the renderer process ID for iframes.
+<pre>
+                           main_frame  [depth:0]
+                         /     |       \
+      first_level_iframe_0     ...    first_level_iframe_n             [depth:1, first level iframe]
+                |          \                     |
+      second_level_iframe_0 ...       second_level_iframe_n ...        [depth:2, second level iframe]
+               |
+              ...                                                      
+</pre> 
+
+In this document we describe the updated API. We'd appreciate your feedback.
+
+# Description
+We propose extending `CoreWebView2` and `CoreWebView2Frame` to include the 
+`RendererProcessId` property and  `RendererProcessIdChanged` event. This property 
+stores the renderer process ID for its iframe. This event will be raised whenever 
+the renderer process ID is changed for `CoreWebView2` or `CoreWebView2Frame`.
+
+Additionally, we propose to add the `FrameCreated` event to `CoreWebView2Frame`. 
+This event is raised when a new iframe is created.
+
+# Examples
+C++
+```c++
+void InitializeEventView(wil::com_ptr<ICoreWebView2> webview)
+{
+    auto webview2_18 = webview.try_query<ICoreWebView2_18>();
+    if (webview2_18) 
+    {
+        webview2_18->add_FrameCreated(
+            Callback<ICoreWebView2FrameCreatedEventHandler>(
+                [this](ICoreWebView2* sender, ICoreWebView2FrameCreatedEventArgs* args)
+                    -> HRESULT {
+                    // ...
+                    // The new frame created event can be recursively initialized
+                    InitializeFrameEventView(webviewFrame);
+                    // ...
+                    return S_OK;
+                })
+                .Get(),
+            NULL);
+        //! [ICoreWebView2:add_RendererProcessIdChanged]
+        webview2_18->add_RendererProcessIdChanged(
+            Callback<ICoreWebView2RendererProcessIdChangedEventHandler>(
+                [this](ICoreWebView2* sender, IUnknown* args) -> HRESULT
+                {
+                    wil::com_ptr<ICoreWebView2_18> webview2_18;
+                    sender->QueryInterface(IID_PPV_ARGS(&webview2_18));
+                    UINT32 processId;
+                    //! [ICoreWebView2:get_RendererProcessId]
+                    CHECK_FAILURE(webview2_18->get_RendererProcessId(&processId));
+                    // Do something when renderer process id has been changed. 
+                    // For instance, retrieve the updated memory usage information 
+                    // for that main frame. 
+                    return S_OK;
+                })
+                .Get(),
+            NULL);
+    }
+}
+
+void InitializeFrameEventView(

```

Comment:
This event handler seems to be same as above so far. The actual scenario her would be to build a hierarchy of frames in order to be able to sum up the memory per top level frame?

Reply:
Correct it will build a hierarchy of frames. Using this recursive method to provide an example of how to capture all iframes for a page. If we don't want to catch all level, we can also use iterative way to only capture to certain level.

Reply:
No change required. We have to have different event handlers because the source types are different.